scriptworker/cot/verify.py

```diff
@@ -0,0 +1,1141 @@
+        )
+        try:
+            with open(path, "r") as fh:
+                contents = fh.read()
+        except OSError as exc:
+            raise CoTError("Can't read {}: {}!".format(path, str(exc)))
+        try:
+            # TODO remove verify_sig pref and kwarg when git repo pubkey
+            # verification works reliably!
+            body = get_body(
+                gpg, contents,
+                verify_sig=chain.context.config['verify_cot_signature']
+            )
+        except ScriptWorkerGPGException as exc:
+            raise CoTError("GPG Error verifying chain of trust for {}: {}!".format(path, str(exc)))
+        link.cot = load_json(
+            body, exception=CoTError,
+            message="{} {}: Invalid cot json body! %(exc)s".format(link.name, link.task_id)
+        )
+        unsigned_path = os.path.join(link.cot_dir, 'chainOfTrust.json')
+        with open(unsigned_path, "w") as fh:
+            fh.write(format_json(link.cot))
+
+
+# verify_link_in_task_graph {{{1
+def verify_link_in_task_graph(chain, decision_link, task_link):
+    """Compare the runtime task definition against the decision task graph.
+
+    Args:
+        chain (ChainOfTrust): the chain we're operating on.
+        decision_link (LinkOfTrust):
+
+    Raises:
+        CoTError: on failure.
+    """
+    log.info("Verifying the {} {} task definition is part of the {} {} task graph...".format(
+        task_link.name, task_link.task_id, decision_link.name, decision_link.task_id
+    ))
+    ignore_keys = ("created", "deadline", "expires", "dependencies", "schedulerId")
+    errors = []
+    runtime_defn = deepcopy(task_link.task)
+    graph_defn = deepcopy(decision_link.task_graph[task_link.task_id])
+    # dependencies
+    bad_deps = set(graph_defn['task']['dependencies']).symmetric_difference(set(runtime_defn['dependencies']))
+    if bad_deps and runtime_defn['dependencies'] != [task_link.decision_task_id]:
+        errors.append("{} {} dependencies don't line up!\n{}".format(
+            task_link.name, task_link.task_id, bad_deps
+        ))
+    # payload - eliminate the 'expires' key from artifacts because the datestring
+    # will change
+    for payload in (runtime_defn['payload'], graph_defn['task']['payload']):
+        for key, value in payload.get('artifacts', {}).items():
+            if isinstance(value, dict) and 'expires' in value:
+                del(value['expires'])
+    # test all non-ignored key/value pairs in the task defn
+    for key, value in graph_defn['task'].items():
+        if key in ignore_keys:
+            continue
+        if value != runtime_defn[key]:
+            errors.append("{} {} {} differs!\n graph: {}\n task: {}".format(
+                task_link.name, task_link.task_id, key,
+                pprint.pformat(value), pprint.pformat(runtime_defn[key])
+            ))
+    raise_on_errors(errors)
+
+
+# verify_firefox_decision_command {{{1
+def verify_firefox_decision_command(decision_link):
```

Comment:
Yes.  I agree 100%.

1. Enable chain of trust verification for dep signing jobs, and make them visible in treeherder.  That way chain of trust bustage is viewable per-push rather than waiting til nightlies or release
2. Put in these rules, and then work towards simplifying them.  Catlee and I were talking and we're thinking we can simplify decision task command logic if we put more logic into the docker image.

But as is, we have to have rules around the decision task, otherwise anyone with the scopes can launch any decision task with any arbitrary env and commandline, which can result in any arbitrary graph.